kotlinx-coroutines-core/common/src/CoroutineStart.kt

```diff
@@ -1,74 +1,342 @@
+     *             }
+     *         }
+     *     }
+     * } finally {
+     *     withContext(NonCancellable) {
+     *         repeat(N_PERMITS) { semaphore.acquire() }
+     *         println("All permits were successfully returned!")
+     *     }
+     * }
+     * ```
+     *
+     * Here, we used [ATOMIC] to ensure that a semaphore that was acquired outside of the coroutine does get released
+     * even if cancellation happens between `acquire()` and `launch`.
+     * As a result, the semaphore will eventually regain all three permits.
+     *
+     * Behavior of [ATOMIC] can be described with the following examples:
+     *
+     * ```
+     * // Example of cancelling atomically started coroutines
+     * runBlocking {
+     *     println("1. Atomically starting a coroutine that goes through a dispatch.")
+     *     launch(start = CoroutineStart.ATOMIC) {
+     *         check(!isActive) // attempting to suspend will throw
+     *         println("4. A coroutine that went through a dispatch also starts.")
+     *         try {
+     *             delay(10.milliseconds)
+     *             println("This code will never run.")
+     *         } catch (e: CancellationException) {
+     *             println("5. Cancellation at later points still works.")
+     *             throw e
+     *         }
+     *     }
+     *     println("2. Cancelling this coroutine and all of its children.")
+     *     cancel()
+     *     launch(Dispatchers.Unconfined, start = CoroutineStart.ATOMIC) {
+     *         check(!isActive) // attempting to suspend will throw
+     *         println("3. An undispatched coroutine starts.")
+     *     }
+     *     ensureActive() // we can even crash the current coroutine.
+     * }
+     * ```
      */
     @ExperimentalCoroutinesApi // Since 1.0.0, no ETA on stability
     ATOMIC,
 
     /**
-     * Immediately executes the coroutine until its first suspension point _in the current thread_ similarly to
-     * the coroutine being started using [Dispatchers.Unconfined]. However, when the coroutine is resumed from suspension
-     * it is dispatched according to the [CoroutineDispatcher] in its context.
+     * Immediately executes the coroutine until its first suspension point _in the current thread_.
+     *
+     * Starting a coroutine using [UNDISPATCHED] is similar to using [Dispatchers.Unconfined] with [DEFAULT], except:
+     * - Resumptions from later suspensions will properly use the actual dispatcher from the coroutine's context.
+     *   Only the code until the first suspension point will be executed immediately.
+     * - Even if the coroutine was cancelled already, its code will still start to be executed, similar to [ATOMIC].
+     *
+     * This set of behaviors makes [UNDISPATCHED] well-suited for cases where the coroutine has a distinct
+     * initialization phase whose side effects we want to rely on later.
+     *
+     * Example:
+     * ```
+     * runBlocking {
```

Comment:
Let's omit this example and leave only the trivial one?

Reply:
Good idea! Used as is.